### ISSUE TITLE
Tuple variant cast as usize is very likely not what you want

### ISSUE THREAD

**Issue by GuillaumeGomez** [rust-lang/rust-clippy#11615]
### What it does

```rust
enum UwU {
    A(String),
    B,
}

fn main() {
    println!("{}", UwU::A as usize);
}
```

In such case, `Uwu::A as usize` will return the function pointer address, which is very likely not what the user intended to do.

### Drawbacks

Maybe it's actually something that the user wants?

### Example

It should suggest to not do that.

**Comment by obi1kenobi** [comment]
I suggested to @GuillaumeGomez that this should be a lint, based on this example that IMHO has even more "not what you want" energy:

```rust
#[repr(usize)]
enum UwU {
    A(String),
    B,
}

fn main() {
    println!("{}", UwU::A as usize);  // works fine, but does something completely unexpected

    UwU::B as usize;  // fails to compile, likely confusingly so since the enum is `#[repr(usize)]` and `UwU::A as usize` compiled fine
}
```

**Comment by Centri3** [comment]
The second example is probably best in rustc, i.e., adding a note that you can't cast a variant of a non-unit enum to a `usize` to get its discriminant, or smth like that.

clippy can't really do anything more than what it already does as I don't think linting occurs if typeck fails, and I don't think there's a way to lint after name resolution but before typeck (since we need the resolver for this)
